### ISSUE TITLE
vtable detection for VC++ binaries

### ISSUE THREAD

**Issue by pinkflawd** [radare/radare2#6851]
Attached binary is malware written in C++, lots of vtables, not detected with av command.
password is infected 

[banito.zip](https://github.com/radare/radare2/files/802015/banito.zip)

**Comment by Maijin** [comment]
Code is here https://github.com/radare/radare2/blob/master/libr/core/anal_vt.c

**Comment by Maijin** [comment]
See implementation here https://github.com/REhints/HexRaysCodeXplorer

![image](https://cloud.githubusercontent.com/assets/1408600/23463396/969154c4-fe91-11e6-8c33-cd3b2d73e26a.png)

**Comment by pinkflawd** [comment]
This blog describes how MS VC++ works quite well http://www.openrce.org/articles/full_view/23

The problem with MS VC++ binaries is, that they don't necessarily come with information on their class structure, there is just something called RTTI, which, at least for malware, is almost always stripped. Finding vtables within the binary is done either sweeping the code section for .. well things that look like vtable structures, or, as done by CodeExplorer, starting from the code and searching for constructors and hoping to find a vtable offset within the arguments.
Neat would be, trying to recover not only vtables, but entire object structures. This would require lots of constructor detection, parsing, and I think guessing, though. 

I'm sure there are more thorough ways of doing this, will think about it and update this thread.

**Comment by XVilka** [comment]
I'm renaming this issue in the [META] for all vtables and C++ metainformation.

**Comment by XVilka** [comment]
@codeuchiha for your reference

**Comment by PankajKataria** [comment]
Thank you @XVilka @pinkflawd @Maijin for providing resources, however I am stuck at parsing RTTI for GCC it will very helpful if some one can point out resources for that too.

**Comment by pinkflawd** [comment]
Cool, thanks for the link! 
As for GCC RTTI or RTTI in general I know this presentation http://www.hexblog.com/wp-content/uploads/2012/06/Recon-2012-Skochinsky-Compiler-Internals.pdf - contains some info. A note though, not sure how it is with benign binaries, but malware doesn't usually come with RTTI information.

**Comment by PankajKataria** [comment]
@pinkflawd I totally agree that the binary doesn't always comes with RTTI structures, but we are first aiming to develop for the case where RTTI is present and then will keep improving including different scenarios.

**Comment by pinkflawd** [comment]
sure sure :)

**Comment by XVilka** [comment]
Also a scripts from our lovely Binary Ninja https://github.com/trailofbits/binjascripts/tree/master/vtable-navigator

**Comment by XVilka** [comment]
Adding one more for vtables/C++ metainfo parsing - https://github.com/igogo-x86/HexRaysPyTools

**Comment by awhawks** [comment]
This from REcon 2011 look useful as well - [Practical C++ Decompilation](http://www.hexblog.com/wp-content/uploads/2011/08/Recon-2011-Skochinsky.pdf)

**Comment by XVilka** [comment]
See also https://llvm.org/docs/HowToSetUpLLVMStyleRTTI.html

**Comment by XVilka** [comment]
This one for Swift https://mikeash.com/pyblog/friday-qa-2014-07-18-exploring-swift-memory-layout.html

**Comment by XVilka** [comment]
This one is for ObjectiveC http://cocoasamurai.blogspot.com/2010/01/understanding-objective-c-runtime.html

**Comment by XVilka** [comment]
New plugin by NCC Group https://github.com/nccgroup/PythonClassInformer
 https://www.nccgroup.trust/us/about-us/newsroom-and-events/blog/2017/october/python-class-informer-an-idapython-plugin-for-viewing-run-time-type-information-rtti/

**Comment by XVilka** [comment]
See also vtables/RTTI parsing from RetDec https://github.com/avast-tl/retdec/tree/master/src/bin2llvmir/optimizations/vtable

**Comment by radare** [comment]
See https://github.com/bazad/memctl

**Comment by XVilka** [comment]
Also https://github.com/cocoahuke/maclook4ref

**Comment by XVilka** [comment]
See also https://github.com/RUB-SysSec/Marx thanks to @radare 
Paper is here https://www.syssec.rub.de/media/emma/veroeffentlichungen/2016/12/22/marx_ndss2017.pdf

**Comment by Maijin** [comment]
https://github.com/whitequark/binja_itanium_cxx_abi

**Comment by XVilka** [comment]
Future implementers - note, that current code in libr/core/anal_vt.c is flawed since it targets x86 only. The detection should be crossplatform.

**Comment by XVilka** [comment]
See also https://alschwalm.com/blog/static/2016/12/17/reversing-c-virtual-functions/

**Comment by XVilka** [comment]
See also more in https://www.blackhat.com/presentations/bh-dc-07/Sabanal_Yason/Paper/bh-dc-07-Sabanal_Yason-WP.pdf

**Comment by XVilka** [comment]
And also http://blog.httrack.com/blog/2014/05/09/a-basic-glance-at-the-virtual-table/ for GCC

**Comment by thestr4ng3r** [comment]
I will try to do msvc soon.

**Comment by Maijin** [comment]
@thestr4ng3r @XVilka @r00tus3r can you check the boxes in the initial post of the issue so we know what is remaining to do here?

**Comment by XVilka** [comment]
@sivaramaaa I think along with types information loading from PDB and DWARF we can load C++ classes as structures at first. Assigning you to think about how it can be done the fastest/easiest way.

@Maijin done, we need to add more tests into r2r for this.

**Comment by XVilka** [comment]
Added this paper https://www.cs.tau.ac.il/~maon/pubs/2018-asplos.pdf

**Comment by XVilka** [comment]
See also (Pharos) https://edmcman.github.io/papers/ccs18.pdf + https://edmcman.github.io/pres/ccs18.pdf

**Comment by XVilka** [comment]
DeClassifier: Class-Inheritance InferenceEngine for Optimized C++ Binaries
[1901.10073.pdf](https://github.com/radare/radare2/files/2836979/1901.10073.pdf)

**Comment by XVilka** [comment]
One more implementation: https://github.com/astrelsky/Ghidra-Cpp-Class-Analyzer

**Comment by ret2libc** [comment]
This issue has been moved from radareorg/radare2 to radareorg/ideas as we are trying to clean our backlog and this issue has probably been created a long while ago. This is an effort to help contributors understand what are the actionable items they can work on, prioritize issues better and help users find active/duplicated issues more easily. If this is not an enhancement/improvement/general idea but a bug, feel free to ask for re-transfer to main repo. Thanks for your understanding and contribution with this issue.
